commitlint.config.js

```diff
@@ -0,0 +1,56 @@
+// commitlint.config.js
+module.exports = {
+    rules: {
+        // rules are defined as [ severity (2=error), "always"|"never", argument? ]. See
+        // https://commitlint.js.org/#/reference-rules?id=rules
+
+        // log-commit-info is used for debugging. It outputs the dict passed to a plugin.
+        // 'log-commit-info': [2, 'always'],
+        'breaking-change-indicator': [2, 'always'],
+        'type-enum': [2, 'always',
+            [
+                'feat',
+                'fix',
+                'dep',
+                'ci',
+                'refactor',
+                'doc',
+                'test',
+            ]
+        ]
+    },
+    parserPreset: {
+        parserOpts: {
+            // add matching groups for `!` before and after scope. Before scope is invalid, after scope is correct
+            // and enforces the "breaking-change" header (see below).
+            headerPattern: /^(\w*)(!?)(?:\(([\w\$\.\-\* ]*)\))?(!?)\: (.*)$/,
+            headerCorrespondence: ["type", "wrong_breaking", "scope", "breaking", "subject"],
+        },
+    },
+    plugins: [
+        {
+            rules: {
+                // log-commit-info is used for debugging. It outputs the dict passed to a plugin.
+                // 'log-commit-info': (input) => {
+                //     console.log(input)
+                //     return [true, "see above"]
+                // },
+                'breaking-change-indicator': ({wrong_breaking, breaking, scope, body, footer}) => {
+                    if (scope && wrong_breaking) {
+                        return [
+                            false, "The breaking change indicator `!` can only be used right before `:`, e.g. `feat(scope)!:`."
+                        ]
+                    }
+
+                    let rxBreakingChange = /^breaking(\s|-)change: .+/img
+
+                    // the marker "breaking change:" is treated as footer and "note" by commitlint, and not considered part of the body.
+                    return [
+                        !breaking || body?.match(rxBreakingChange) || footer?.match(rxBreakingChange),
```

Comment:
writing this a bit more verbously wouldn't hurt but also fine as is.

Reply:
Pulled out to a  variable and added explanations for the individually ORed sub-expressions